transcendence-app/src/game/game.queue.service.ts

```diff
@@ -0,0 +1,146 @@
+import { Injectable } from '@nestjs/common';
+import { Server } from 'socket.io';
+import { GameId, UserId } from './infrastructure/db/memoryModels';
+import { IGamesOngoingRepository } from './infrastructure/db/gamesongoing.repository';
+import { IChallengesPendingRepository } from './infrastructure/db/challengespending.repository';
+import {
+  GameChallengeDto,
+  GameChallengeResponseDto,
+  gameQueueServerToClientWsEvents,
+} from 'pong-engine';
+import { GamePairingStatusDto } from './dto/game.pairing.status.dto';
+
+@Injectable()
+export class GameQueueService {
+  public socket: Server | null = null;
+  private gameQueue: Record<GameId, [UserId, null]> | null = null;
+  constructor(
+    private gamesOngoing: IGamesOngoingRepository,
+    private challengesPending: IChallengesPendingRepository,
+  ) {}
+
+  private getWaitingUserId(): string | null {
+    if (!this.gameQueue) return null;
+    return Object.values(this.gameQueue)[0][0];
+  }
+
+  private getWaitingGameId(): string | null {
+    if (!this.gameQueue) return null;
+    return Object.keys(this.gameQueue)[0];
+  }
+
+  private isUserBusy(userId: string): boolean {
+    return (
+      this.gamesOngoing.isPlayerBusy(userId) ||
+      this.challengesPending.isPlayerBusy(userId) ||
+      this.getWaitingUserId() === userId
+    );
+  }
+
+  async gameQueueJoin(userId: string): Promise<[string, boolean] | null> {
+    if (!this.socket) {
+      return null;
+    }
+    if (this.isUserBusy(userId)) {
+      // Noop, we don't want to kick the player from anywhere
+      // early noop in case the server does know that the user is busy
+      return null;
+    }
+    if (!this.gameQueue) {
+      if (this.getWaitingUserId() === userId) {
+        return null;
+      }
+      this.gameQueue = this.gamesOngoing.addGame(userId);
+      const gameRoomId = this.getWaitingGameId();
+      if (!gameRoomId) return null;
```

Comment:
We should never have this condition be succesful, since having a user in the queue should mean that it's a assigned a gameRoomId, but the ts-linter does not allow me to make type assertions, and working around it it's harder than placing an if here. But yes, I guess, if there is a bug and we do hit this block, I should at least throw an exception.

Reply:
With the new type for the queue, now this is not possible and the linter does not require it. Done.